File: backend/app/services/comparator.py
import re
from typing import Optional, Tuple
from ..models.schemas import (
    ParameterRequirement,
    DeviceParameter,
    ComparisonResult,
    ComplianceStatus
)


class ParameterComparator:
    """Serwis do porównywania parametrów urządzeń z wymaganiami minimalnymi"""

    def compare(
        self,
        requirement: ParameterRequirement,
        device_param: Optional[DeviceParameter]
    ) -> ComparisonResult:
        """Porównuje parametr urządzenia z wymaganiem minimalnym"""

        if not device_param:
            return ComparisonResult(
                parametr=requirement.nazwa,
                wymaganie_minimalne=requirement.wymaganie_minimalne,
                wartosc_urzadzenia="BRAK DANYCH",
                status=ComplianceStatus.DO_WERYFIKACJI,
                komentarz="Nie znaleziono tego parametru w karcie katalogowej"
            )

        req_value = requirement.wymaganie_minimalne
        dev_value = device_param.wartosc
        comparison_type = requirement.typ_porownania or "exact"

        # Wybierz metodę porównania
        if comparison_type == "min":
            status, comment = self._compare_min(req_value, dev_value)
        elif comparison_type == "max":
            status, comment = self._compare_max(req_value, dev_value)
        elif comparison_type == "boolean":
            status, comment = self._compare_boolean(req_value, dev_value)
        elif comparison_type == "contains":
            status, comment = self._compare_contains(req_value, dev_value)
        else:
            status, comment = self._compare_exact(req_value, dev_value)

        return ComparisonResult(
            parametr=requirement.nazwa,
            wymaganie_minimalne=req_value,
            wartosc_urzadzenia=dev_value,
            status=status,
            komentarz=comment
        )

    def _extract_number(self, text: str) -> Optional[float]:
        """Wyciąga liczbę z tekstu"""
        # Usuń spacje i zamień przecinek na kropkę
        text = text.replace(" ", "").replace(",", ".")

        # Szukaj liczb (w tym z jednostkami)
        patterns = [
            r'(\d+\.?\d*)',  # podstawowa liczba
            r'(\d+)',  # liczba całkowita
        ]

        for pattern in patterns:
            match = re.search(pattern, text)
            if match:
                try:
                    return float(match.group(1))
                except ValueError:
                    continue

        return None

    def _compare_min(self, req: str, dev: str) -> Tuple[ComplianceStatus, str]:
        """Porównanie minimalne (wartość urządzenia >= wymaganie)"""
        req_num = self._extract_number(req)
        dev_num = self._extract_number(dev)

        if req_num is None or dev_num is None:
            return ComplianceStatus.DO_WERYFIKACJI, "Nie można porównać wartości numerycznych"

        if dev_num >= req_num:
            return ComplianceStatus.SPEŁNIA, f"{dev_num} >= {req_num}"
        else:
            return ComplianceStatus.NIE_SPEŁNIA, f"{dev_num} < {req_num} (wymagane minimum)"

    def _compare_max(self, req: str, dev: str) -> Tuple[ComplianceStatus, str]:
        """Porównanie maksymalne (wartość urządzenia <= wymaganie)"""
        req_num = self._extract_number(req)
        dev_num = self._extract_number(dev)

        if req_num is None or dev_num is None:
            return ComplianceStatus.DO_WERYFIKACJI, "Nie można porównać wartości numerycznych"

        if dev_num <= req_num:
            return ComplianceStatus.SPEŁNIA, f"{dev_num} <= {req_num}"
        else:
            return ComplianceStatus.NIE_SPEŁNIA, f"{dev_num} > {req_num} (wymagane maksimum)"

    def _compare_boolean(self, req: str, dev: str) -> Tuple[ComplianceStatus, str]:
        """Porównanie tak/nie"""
        positive_words = ["tak", "yes", "true", "1", "posiada", "obsługuje", "wspiera", "✓", "✔"]
        negative_words = ["nie", "no", "false", "0", "brak", "✗", "✘", "-"]

        req_lower = req.lower().strip()
        dev_lower = dev.lower().strip()

        # Jeśli wymaganie to "tak"
        if any(word in req_lower for word in positive_words):
            if any(word in dev_lower for word in positive_words):
                return ComplianceStatus.SPEŁNIA, "Urządzenie posiada wymaganą funkcję"
            elif any(word in dev_lower for word in negative_words):
                return ComplianceStatus.NIE_SPEŁNIA, "Urządzenie nie posiada wymaganej funkcji"

        return ComplianceStatus.DO_WERYFIKACJI, "Wymaga manualnej weryfikacji"

    def _compare_contains(self, req: str, dev: str) -> Tuple[ComplianceStatus, str]:
        """Sprawdza czy urządzenie zawiera wymaganą funkcjonalność"""
        req_lower = req.lower()
        dev_lower = dev.lower()

        # Wyciągnij kluczowe słowa z wymagania
        keywords = self._extract_keywords(req_lower)

        matched = sum(1 for kw in keywords if kw in dev_lower)
        total = len(keywords)

        if total == 0:
            return ComplianceStatus.DO_WERYFIKACJI, "Brak słów kluczowych do porównania"

        if matched == total:
            return ComplianceStatus.SPEŁNIA, f"Znaleziono wszystkie wymagane elementy ({matched}/{total})"
        elif matched > 0:
            return ComplianceStatus.DO_WERYFIKACJI, f"Częściowe dopasowanie ({matched}/{total})"
        else:
            return ComplianceStatus.NIE_SPEŁNIA, "Nie znaleziono wymaganych elementów"

    def _compare_exact(self, req: str, dev: str) -> Tuple[ComplianceStatus, str]:
        """Porównanie dokładne lub przybliżone"""
        req_clean = req.lower().strip()
        dev_clean = dev.lower().strip()

        # Dokładne dopasowanie
        if req_clean == dev_clean:
            return ComplianceStatus.SPEŁNIA, "Dokładne dopasowanie"

        # Sprawdź czy wymaganie zawiera się w wartości urządzenia
        if req_clean in dev_clean:
            return ComplianceStatus.SPEŁNIA, "Wartość urządzenia zawiera wymaganie"

        # Porównaj liczby jeśli są
        req_num = self._extract_number(req)
        dev_num = self._extract_number(dev)

        if req_num is not None and dev_num is not None:
            if dev_num >= req_num:
                return ComplianceStatus.SPEŁNIA, f"Wartość {dev_num} >= {req_num}"
            else:
                return ComplianceStatus.NIE_SPEŁNIA, f"Wartość {dev_num} < {req_num}"

        # Sprawdź podobieństwo tekstu
        similarity = self._text_similarity(req_clean, dev_clean)
        if similarity > 0.8:
            return ComplianceStatus.SPEŁNIA, f"Wysokie podobieństwo ({similarity:.0%})"
        elif similarity > 0.5:
            return ComplianceStatus.DO_WERYFIKACJI, f"Częściowe podobieństwo ({similarity:.0%})"

        return ComplianceStatus.DO_WERYFIKACJI, "Wymaga manualnej weryfikacji"

    def _extract_keywords(self, text: str) -> list:
        """Wyciąga kluczowe słowa z tekstu"""
        # Usuń słowa funkcyjne
        stopwords = {"i", "lub", "oraz", "a", "w", "z", "do", "na", "dla", "min", "max", "co", "najmniej"}
        words = re.findall(r'\b\w+\b', text.lower())
        return [w for w in words if w not in stopwords and len(w) > 2]

    def _text_similarity(self, text1: str, text2: str) -> float:
        """Oblicza podobieństwo dwóch tekstów (Jaccard)"""
        words1 = set(text1.split())
        words2 = set(text2.split())

        if not words1 or not words2:
            return 0.0

        intersection = words1 & words2
        union = words1 | words2

        return len(intersection) / len(union)

File: src/lib/comparator.ts
import {
  ParameterRequirement,
  DeviceParameter,
  ComparisonResult,
  ComplianceStatus
} from '@/types'
import { extractNumber } from './pdf-parser'

export function compareParameter(
  requirement: ParameterRequirement,
  deviceParam: DeviceParameter | null
): ComparisonResult {
  if (!deviceParam) {
    return {
      parametr: requirement.nazwa,
      wymaganie_minimalne: requirement.wymaganie_minimalne,
      wartosc_urzadzenia: 'BRAK DANYCH',
      status: 'DO WERYFIKACJI',
      komentarz: 'Nie znaleziono tego parametru w karcie katalogowej'
    }
  }

  const reqValue = requirement.wymaganie_minimalne
  const devValue = deviceParam.wartosc
  const comparisonType = requirement.typ_porownania || 'exact'

  let status: ComplianceStatus
  let komentarz: string

  switch (comparisonType) {
    case 'min':
      [status, komentarz] = compareMin(reqValue, devValue)
      break
    case 'max':
      [status, komentarz] = compareMax(reqValue, devValue)
      break
    case 'boolean':
      [status, komentarz] = compareBoolean(reqValue, devValue)
      break
    case 'contains':
      [status, komentarz] = compareContains(reqValue, devValue)
      break
    default:
      [status, komentarz] = compareExact(reqValue, devValue)
  }

  return {
    parametr: requirement.nazwa,
    wymaganie_minimalne: reqValue,
    wartosc_urzadzenia: devValue,
    status,
    komentarz
  }
}

function compareMin(req: string, dev: string): [ComplianceStatus, string] {
  const reqNum = extractNumber(req)
  const devNum = extractNumber(dev)

  if (reqNum === null || devNum === null) {
    return ['DO WERYFIKACJI', 'Nie można porównać wartości numerycznych']
  }

  if (devNum >= reqNum) {
    return ['SPEŁNIA', `${devNum} >= ${reqNum}`]
  }
  return ['NIE SPEŁNIA', `${devNum} < ${reqNum} (wymagane minimum)`]
}

function compareMax(req: string, dev: string): [ComplianceStatus, string] {
  const reqNum = extractNumber(req)
  const devNum = extractNumber(dev)

  if (reqNum === null || devNum === null) {
    return ['DO WERYFIKACJI', 'Nie można porównać wartości numerycznych']
  }

  if (devNum <= reqNum) {
    return ['SPEŁNIA', `${devNum} <= ${reqNum}`]
  }
  return ['NIE SPEŁNIA', `${devNum} > ${reqNum} (wymagane maksimum)`]
}

function compareBoolean(req: string, dev: string): [ComplianceStatus, string] {
  const positiveWords = ['tak', 'yes', 'true', '1', 'posiada', 'obsługuje', 'wspiera', '✓', '✔']
  const negativeWords = ['nie', 'no', 'false', '0', 'brak', '✗', '✘', '-']

  const reqLower = req.toLowerCase().trim()
  const devLower = dev.toLowerCase().trim()

  // Jeśli wymaganie to "tak"
  if (positiveWords.some(w => reqLower.includes(w))) {
    if (positiveWords.some(w => devLower.includes(w))) {
      return ['SPEŁNIA', 'Urządzenie posiada wymaganą funkcję']
    }
    if (negativeWords.some(w => devLower.includes(w))) {
      return ['NIE SPEŁNIA', 'Urządzenie nie posiada wymaganej funkcji']
    }
  }

  return ['DO WERYFIKACJI', 'Wymaga manualnej weryfikacji']
}

function compareContains(req: string, dev: string): [ComplianceStatus, string] {
  const reqLower = req.toLowerCase()
  const devLower = dev.toLowerCase()

  const keywords = extractKeywords(reqLower)
  const matched = keywords.filter(kw => devLower.includes(kw)).length
  const total = keywords.length

  if (total === 0) {
    return ['DO WERYFIKACJI', 'Brak słów kluczowych do porównania']
  }

  if (matched === total) {
    return ['SPEŁNIA', `Znaleziono wszystkie wymagane elementy (${matched}/${total})`]
  }
  if (matched > 0) {
    return ['DO WERYFIKACJI', `Częściowe dopasowanie (${matched}/${total})`]
  }
  return ['NIE SPEŁNIA', 'Nie znaleziono wymaganych elementów']
}

function compareExact(req: string, dev: string): [ComplianceStatus, string] {
  const reqClean = req.toLowerCase().trim()
  const devClean = dev.toLowerCase().trim()

  // Dokładne dopasowanie
  if (reqClean === devClean) {
    return ['SPEŁNIA', 'Dokładne dopasowanie']
  }

  // Wymaganie zawiera się w wartości urządzenia
  if (devClean.includes(reqClean)) {
    return ['SPEŁNIA', 'Wartość urządzenia zawiera wymaganie']
  }

  // Porównaj liczby
  const reqNum = extractNumber(req)
  const devNum = extractNumber(dev)

  if (reqNum !== null && devNum !== null) {
    if (devNum >= reqNum) {
      return ['SPEŁNIA', `Wartość ${devNum} >= ${reqNum}`]
    }
    return ['NIE SPEŁNIA', `Wartość ${devNum} < ${reqNum}`]
  }

  // Podobieństwo tekstu
  const similarity = textSimilarity(reqClean, devClean)
  if (similarity > 0.8) {
    return ['SPEŁNIA', `Wysokie podobieństwo (${Math.round(similarity * 100)}%)`]
  }
  if (similarity > 0.5) {
    return ['DO WERYFIKACJI', `Częściowe podobieństwo (${Math.round(similarity * 100)}%)`]
  }

  return ['DO WERYFIKACJI', 'Wymaga manualnej weryfikacji']
}

function extractKeywords(text: string): string[] {
  const stopwords = new Set(['i', 'lub', 'oraz', 'a', 'w', 'z', 'do', 'na', 'dla', 'min', 'max', 'co', 'najmniej'])
  const words = text.match(/\b\w+\b/g) || []
  return words.filter(w => !stopwords.has(w) && w.length > 2)
}

function textSimilarity(text1: string, text2: string): number {
  const arr1 = text1.split(/\s+/)
  const arr2 = text2.split(/\s+/)
  const words1 = new Set(arr1)
  const words2 = new Set(arr2)

  if (words1.size === 0 || words2.size === 0) return 0

  let intersectionCount = 0
  arr1.forEach(word => {
    if (words2.has(word)) intersectionCount++
  })

  const unionSet = new Set(arr1.concat(arr2))

  return intersectionCount / unionSet.size
}

export function findMatchingParam(
  paramName: string,
  deviceParams: DeviceParameter[]
): DeviceParameter | null {
  const paramLower = paramName.toLowerCase()

  // Dokładne dopasowanie
  let found = deviceParams.find(dp => dp.nazwa.toLowerCase() === paramLower)
  if (found) return found

  // Częściowe dopasowanie
  found = deviceParams.find(dp =>
    paramLower.includes(dp.nazwa.toLowerCase()) ||
    dp.nazwa.toLowerCase().includes(paramLower)
  )
  if (found) return found

  // Dopasowanie słów kluczowych
  const keywords = paramLower.split(/\s+/)
  found = deviceParams.find(dp => {
    const dpWords = dp.nazwa.toLowerCase().split(/\s+/)
    return keywords.some(kw => dpWords.includes(kw))
  })

  return found || null
}
